design/pipeline-triggers.md

```diff
@@ -0,0 +1,369 @@
+# Triggers in pipelines
+
+Triggers are events on which you can start your pipeline run automatically. You can enable triggers on your pipeline by subscribing to both internal and external events.
+
+At high level there are 3 different types of pipeline triggers.
+1. Resource triggers
+2. Schedule triggers
+3. Webhook triggers
+
+## Resource triggers
+You can enable triggers on the resources defined in your pipeline. Resources can be of types pipelines, repositories, containers and packages.
+Triggers are enabled by default on all the resources. However, you can choose to override/disable triggers for each resource.
+
+
+
+### Pipelines
+A new pipeline is triggered automatically whenever a new run of the `pipeline` resource is succesfully completed. See [pipeline resources](pipeline-resources.md#resources-pipelines) for more details.
+
+#### Scenarios
+- I would like to trigger my pipeline when a new ‘helm-package’ artifact is published by ‘Helm-CI’ pipeline that ran on `releases/*` branch.
+- I would like to trigger my pipeline when a new ‘helm-package’ artifact is published and tested as part of Helm-CI pipeline and tagged as 'Production'.
+- I would like to trigger my pipeline when ‘TFS-Update’ pipeline has completed ‘Ring2’ stage.  
+
+#### Schema
+```yaml
+resources:       
+  pipelines:
+  - pipeline: string 
+    source: string  
+    trigger:     # Optional; Triggers are enabled by default.
+      branches:  # branch conditions to filter the events, optional; Defaults to all branches.
+        include: [ string ]  # branches to consider the trigger events, optional; Defaults to all branches.
+        exclude: [ string ]  # branches to discard the trigger events, optional; Defaults to none.
+      stages: [ string ]  # trigger after completion of given stage, optional; Defaults to all stage completion.
+      tags: [ string ]  # tags on which the trigger events are considered, optional; Defaults to any tag or no tag.
+```
+
+#### Examples
+Triggers are enabled by default unless expliciltly opted out. You can disable the triggers on the `pipeline` resource.
+```yaml
+resources:
+  pipelines:
+  - pipeline: SmartHotel
+    source: SmartHotel-CI 
+    trigger: none   
+```
+
+
+You can control which branches get the triggers with a simple syntax.
```

Comment:
I'm (personally) of very mixed opinions on the simple syntax we have. The major downsides are 1) offering multiple right ways to say the same thing and 2) complicating the learning curve for "growing up" to more sophisticated uses. In a case like `self` repo branch include triggers, where that's far and away going to be the most common scenario, the upside of shorter syntax and the fact that it's reasonably unambiguous pays off that debt. But for pipeline triggers (and other types below), are we sure that's true? (In a nutshell: are you sure you want to have two syntaxes here?)

Reply:
Yeah, I am not comfortable with two syntaxes either. There are two reasons for providing simpler syntax:
1. Since we allowed simpler syntax for self repo, user tends to use the same syntax and expects it to work.
2. The simple syntax works for most of the scenarios.

Reply:
I think it is reasonable to require the full syntax when dealing with triggers on resources.  From the perspective of what the customer is trying to do they are already much deeper in than the simple CI scenario where some of the brevity is driven by trying to compete with some of our simpler competitors.

Reply:
Okay, will remove the short syntax and will not mention anywhere in the spec.
Now, what if the user provides the short syntax since it is supported else where. Are we going to invalidate the YAML or does the parser have the support to understand this.

Reply:
It would be invalid YAML.

Reply:
As discussed, removed the short syntax.